analytics_data_api/utils.py

```diff
@@ -84,3 +94,92 @@ def date_range(start_date, end_date, delta=datetime.timedelta(days=1)):
         cur_date += delta
+
+
+def get_course_report_download_details(course_id, report_name):
+    """
+    Determine the path that the report file should be located at,
+    then return metadata sufficient for downloading it.
+    """
+    report_location_template = getattr(settings, 'COURSE_REPORT_FILE_LOCATION_TEMPLATE')
+    report_location = report_location_template.format(
+        course_id=course_id,
+        report_name=report_name
+    )
+    if getattr(settings, 'DEFAULT_FILE_STORAGE', None) != AWS_S3_FILE_STORAGE_CLASS:
+        # At present, we do not support linking from storage other than S3.
+        raise CannotCreateReportDownloadLinkError
+
+    if not default_storage.exists(report_location):
+        raise ReportFileNotFoundError(course_id=course_id, report_name=report_name)
+
+    try:
+        last_modified = default_storage.modified_time(report_location)
+    except (NotImplementedError, AttributeError):
+        last_modified = None
+
+    try:
+        download_size = default_storage.size(report_location)
+    except (NotImplementedError, AttributeError):
+        download_size = None
+
+    # Note that course IDs contain characters that aren't valid for filenames in
+    # Windows. However, we'll trust the client to save them with the appropriate
+    # alternate naming scheme (replace ':' with '_' and so on)
+    download_filename = '{}-{}-{}.csv'.format(
+        course_id,
+        report_name,
+        # We need a date for the filename; if we don't know when it was last modified,
+        # use the current date and time to stamp the filename.
+        (last_modified or datetime.datetime.utcnow()).strftime('%Y%m%dT%H%M%S')
+    ).replace(
+        ':', '-'
+    ).replace(
+        '+', '-'
+    )
+    url, expiration_date = get_file_object_url(report_location, download_filename)
+
+    details = {
+        'course_id': course_id,
+        'report_name': report_name,
+        'download_url': url
+    }
+    # These are all optional items that aren't guaranteed. The URL isn't guaranteed
+    # either, but we'll raise an exception earlier if we don't have it. Currently, support
+    # is only present for S3, which has all these data types, but this future-proofs
+    # us for scenarios where we might have a storage provider that doesn't support them.
+    if last_modified is not None:
+        details.update({'last_modified': last_modified.strftime(ISO_8601_FORMAT_STRING)})
+    if expiration_date is not None:
+        details.update({'expiration_date': expiration_date.strftime(ISO_8601_FORMAT_STRING)})
+    if download_size is not None:
+        details.update({'file_size': download_size})
+    return details
+
+
+def get_file_object_url(filename, download_filename):
+    """
+    Retrieve a download URL for the file, as well as a datetime object
+    indicating when the URL expires.
+
+    We need to pass extra details to the URL method, above and beyond just the
+    file location, to give us what we need.
+    """
```

Comment:
Please add a comment indicating that this logic may need to be changed to support non-S3 django.storage modules.